python/ray/train/_internal/session.py

```diff
@@ -240,9 +251,27 @@ def _report_legacy(self, **kwargs):
         if self.ignore_report:
             return
 
-        kwargs = self._encode_data_fn(self._auto_fill_metrics(kwargs))
+        # Special case: early fail for Torch tensors
+        if "torch" in sys.modules:
+            from ray.air._internal.torch_utils import contains_tensor
+
+            if contains_tensor(kwargs):
+                raise ValueError(
+                    "Passing objects containg Torch tensors as metrics "
+                    "is not supported as it will throw an exception on "
+                    "deserialization. You can either convert the tensors "
+                    "to Python objects or use a `TorchCheckpoint` as the "
+                    "`checkpoint` argument of `ray.air.session.report` to "
+                    "store your Torch objects."
+                )
```

Comment:
should we move this error to `session.report`? Basically keep the error as close to the user as possible.

This logic can probably be encapsulated in some sort of `_validate_metrics` API to detect any bad metrics directly at the report call.

Reply:
We can move it once `train.report` is completely depreceated.